stan/math/rev/core/autodiffstackstorage.hpp

```diff
@@ -9,40 +9,88 @@ namespace math {
- * case. In the threading case we use the defacto standard C++11
- * singleton pattern relying on a function wrapping a static local
- * variable. This standard pattern is expected to be well supported
- * by the major compilers (as its standard), but it does incur some
- * performance penalty.  There has been some discussion on this; see
- * [1] and [2] and the discussions those PRs link to as well.
+ * For performance reasons the singleton is a global static pointer
+ * which is stored as thread local (TLS) if STAN_THREADS is
+ * defined. The use of a pointer is motivated by performance reasons
+ * for the threading case. When a TLS is used then initialization with
+ * a constant expression is required for fast access to the TLS. As
+ * the AD storage struct is non-POD it must be initialized as a
+ * dynamic expression such that compilers will wrap any access to the
+ * TLS by a TLS wrapper function which causes a significant
+ * slow-down. A pointer to the AD storage instance can be initialized
+ * to a compile-time constant expression of nullptr. In this case the
+ * compiler avoids the use of a TLS wrapper function. Furthermore we
+ * use the __thread keyword on compilers which support it. The
+ * __thread keyword is a compiler-specific (gcc, clang, Intel)
+ * extension which requires initialization with a compile time
+ * constant expression. The C++11 keyword thread_local does allow for
+ * constant and dynamic initialization of the TLS. Thus, only the
+ * __thread keyword gurantees that constant initialization and it's
+ * implied speedup, is used.
  *
- * These are thread_local only if the user asks for it with
- * -DSTAN_THREADS. This is primarily because Apple clang compilers
- * before 2016 don't support thread_local and the additional
- * performance cost. We have proposed removing support for those[3],
- * and at that time we should evaluate the performance of a switch to
- * thread_local.  If there is no loss in performance, we can remove
- * this ifdef.
+ * The initialzation of the AD instance is handled by the lifetime of
+ * a AutodiffStackSingleton object. More specifically, the first
+ * instance of the AutodiffStackSingleton object will initialize the
+ * AD instance and take ownership. Thus whenever the first instance of
+ * the AutodiffStackSingleton object gets destructed, the AD tape will
+ * be destructed as well.  Within stan-math the initialization of the
+ * AD instance for the main thread of the program is handled by
+ * instantiating once the singleton once in the
+ * init_chainablestack.hpp file. Whenever STAN_THREADS is defined then
+ * all created child threads must call the init() method of the AD
+ * singleton in order to initialize the TLS if child threads want to
+ * perform AD operations (the initialization in the main process is
+ * already taken care of in any case).
+ *
+ * The design of a globally held (optionally TLS) pointer, which is
```

Comment:
Is the part about the static init order fiasco still true about the code in the way it is now?

Reply:
Yes, it is still true, but at least now with the changes a C++ user can handle it. Assume that a C++ user performs AD during initialization (so before a program reaches main). In that case it is not a given that the code in `init_chainablestack.hpp` is executed first.